thorlcr/master/thmastermain.cpp

```diff
@@ -441,9 +441,59 @@ bool checkClusterRelicateDAFS(IGroup &grp)
 
 static bool auditStartLogged = false;
 
+class CThorEndHandler : public CSimpleInterface, implements IThreaded
+{
+    CThreaded threaded;
+    unsigned timeout = 30000;
+    std::atomic<bool> started{false};
+    std::atomic<bool> stopped{false};
+    Semaphore sem;
+public:
+    CThorEndHandler() : threaded("CThorEndHandler")
+    {
+        threaded.init(this);
+    }
+    ~CThorEndHandler()
+    {
+        stop();
+        threaded.join(timeout);
+    }
+    void start(unsigned timeoutSecs)
+    {
+        bool expected = false;
+        if (started.compare_exchange_strong(expected, true));
+        {
+            timeout = timeoutSecs * 1000; // sem_post and sem_wait are mem_barriers
+            sem.signal();
+        }
+    }
+    void stop()
+    {
+        bool expected = false;
+        if (stopped.compare_exchange_strong(expected, true));
+            sem.signal();
+    }
+    virtual void threadmain() override
+    {
+        // wait to be signalled to start timer
+        sem.wait();
+        if (stopped)
+            return;
+        if (!sem.wait(timeout))
+            raise(SIGKILL);
```

Comment:
Does this compile on Windows, I wonder?

Reply:
I don't think it will, the other similar case (in ControlHandler in this file) has #ifdef Win32 to call TerminateProcess.

Reply:
Would _exit be a good alternative here?
If not, might be worth adding a TerminateSelf or similar to platform.h - so can use here/other places platform independently.